src/conn/conn_open.c

```diff
@@ -80,28 +80,12 @@ __wt_connection_close(WT_CONNECTION_IMPL *conn)
 	F_SET(conn, WT_CONN_CLOSING);
 	WT_FULL_BARRIER();
 
-	/*
-	 * Clear any pending async operations and shut down the async worker
-	 * threads and system before closing LSM.
-	 */
-	WT_TRET(__wt_async_flush(session));
-	WT_TRET(__wt_async_destroy(session));
-
 	/*
 	 * Shut down server threads other than the eviction server, which is
 	 * needed later to close btree handles.  Some of these threads access
 	 * btree handles, so take care in ordering shutdown to make sure they
 	 * exit before files are closed.
 	 */
-	WT_TRET(__wt_lsm_manager_destroy(session));
-
-	/*
-	 * Once the async and LSM threads exit, we shouldn't be opening any
-	 * more files.
-	 */
-	F_SET(conn, WT_CONN_CLOSING_NO_MORE_OPENS);
-	WT_FULL_BARRIER();
```

Comment:
I am concerned about removing these lines from `wt_connection_close`. That function is called on the error path out of `wiredtiger_open` and if those threads were started and we later got an error, we would no longer call the `destroy` functions here. At a minimum, I'd like to see test cases added that (somehow) force an error in `wiredtiger_open` to convince ourselves this is okay.

Reply:
Maybe have a new function `connection_close_early` that calls the `wt_async_destroy` and `wt_lsm_manager_destroy` and sets the flags that can be called in `conn_close` as well as in the `wiredtiger_open` error path?

Reply:
I just reviewed this again, and please tell me if I'm missing something.

I think there are 3 questions we have to answer:

* Do we start either the LSM or async services before `wiredtiger_open()` returns? If so, there's a problem because `wiredtiger_open()` errors call `__wt_connection_close()` to do cleanup, and the proposed change removes LSM and async shutdown from that cleanup.

* Do the LSM or async services care about the `WT_CONN_CLOSING` flag? If so, there's a problem because removing LSM and async shutdown from `__wt_connection_close()` means LSM and async must correctly shut down absent that flag being set.

* Does the code in `__wt_connection_close()`, executed after LSM and async shutdown, ever open files? If so, there's a problem because the proposed change sets the `WT_CONN_CLOSING_NO_MORE_OPENS` flag earlier than before.

I think the answer to all 3 questions is "No", and so the change is safe.

Do you see a path I'm missing?

Reply:
The answer to #1 is yes. At a minimum for async:
`wt_connection_workers -> wt_async_create -> async_start -> wt_thread_create`